orttraining/orttraining/python/training/ortmodule/_io.py

```diff
@@ -348,139 +271,33 @@ def extract_tensor(value):
-            new_user_output = copy.copy(user_output)
-            for key in sorted(user_output):
-                new_user_output[key] = _replace_stub_with_tensor_value(new_user_output[key], outputs, output_idx)
-            user_output = new_user_output
-        else:
-            raise wrap_exception(
-                ORTModuleIOError,
-                TypeError(f"ORTModule does not support the following model output type {type(user_output)}."),
-            )
-
-        return user_output
-
-    # It is expected that the outputs are ordered in the way defined in the exported onnx model
-    # which is the order in which the output schema was saved.
-    output_idx = [0]
-    user_output = _replace_stub_with_tensor_value(output_schema, outputs, output_idx)
-    return user_output
-
-
-def _extract_schema(data: _ModelInputOutputType, logger: Logger) -> _ModelInputOutputSchemaType:
-    """Extract the data schema by replacing every torch.Tensor value with _TensorStub.
-
-    Depth first traversal to iterate over the data:
-    > Replace every tensor with a stub
-    > Replace None/str typed data with itself
-    > Recreate tensor from data for other primitive types, and replace them with a stub.
-
-    Examples:
-        Example 1, list:
-            data = [torch.tensor(1), torch.tensor(2)]
-            schema = [_TensorStub(shape=()), _TensorStub(shape=())]
-
-        Example 2, dict:
-            data = {"a": torch.tensor(1), "b": torch.tensor(2)}
-            schema = {"a": _TensorStub(shape=()), "b": _TensorStub(shape=())}
-
-        Example 3, dict of list:
-            data = {"a": [torch.tensor(1), torch.tensor(2)], "b": [torch.tensor(3), torch.tensor(4)]}
-            schema = {"a": [_TensorStub(shape=()), _TensorStub(shape=())],
-                        "b": [_TensorStub(shape=()), _TensorStub(shape=())]}
-
-        Example 4, nested dict:
-            data = {"a": {"b": torch.tensor(1), "c": torch.tensor(2)},
-                    "d": {"e": torch.tensor(3), "f": torch.tensor(4)}}
-            schema = {"a": {"b": _TensorStub(shape=()), "c": _TensorStub(shape=())},
-                        "d": {"e": _TensorStub(shape=()), "f": _TensorStub(shape=())}}
-
-        Example 5, dict of mixed list and dict:
-            data = {"a": [torch.tensor(1), torch.tensor(2)], "b": {"c": torch.tensor(3), "d": torch.tensor(4)}}
-            schema = {"a": [_TensorStub(shape=()), _TensorStub(shape=())],
-                        "b": {"c": _TensorStub(shape=()), "d": _TensorStub(shape=())}}
-
-    Args:
-        data: The data to extract the schema from, which can be in any kind of nested structure,
-            including Sequence and Mapping.
-
-
-    Returns:
-        The schema of the data, which has the same structure as the data.
+def unflatten_user_output(output_schema: Optional[ORTModelInputOutputSchemaType], outputs: List[torch.Tensor]):
+    try:
+        return unflatten_from_data_and_schema(outputs, output_schema)
```

Comment:
Can we call this function `unflatten_data_using_schema` or `unflatten_data_with_schema`?

Otherwise, it sounds like both the data and schema are being unflattenend.

Reply:
Indeed. Changed names  to extract_data_and_schema and unflatten_data_using_schema